### PR TITLE
fix(parse): suggest `--` when a refused flag could have been a value

### PR DESCRIPTION
**Pull request by jdx**
`mise exec node --version` answered with a tip that pointed the wrong way:

```
error: unexpected argument '--version' found

  tip: a similar argument exists: '--verbose'

Usage: mise exec [FLAGS] [TOOL@VERSION]… [-- COMMAND]…
```

The user did not want `--verbose`. They wanted `--version` handed to the command
`exec` runs, and the `--` that does that went unmentioned — even though `exec`
declares a `double_dash = "required"` positional and so cannot take a value any
other way. After this change:

```
error: unexpected argument '--version' found

  tip: a similar argument exists: '--verbose'
  tip: to pass '--version' as a value, use '-- --version'

Usage: mise exec [FLAGS] [TOOL@VERSION]… [-- COMMAND]…
```

## The rule

clap prints this tip, so this is clap's rule transcribed rather than reinvented —
its exception included. From `clap_builder/src/parser/parser.rs`:

```rust
// `did_you_mean` is a lot more likely and should cause us to skip the `--` suggestion
// with the one exception being that the CLI is trying to capture arguments
let suggested_trailing_arg = (did_you_mean.is_none()
    || self.cmd.get_positionals().any(|arg| arg.is_last_set() || arg.is_trailing_var_arg_set()))
    && !trailing_values
    && self.cmd.has_positionals();
```

In usage's vocabulary: the command takes positionals, no `--` already stands
earlier on the line, and either there was no spelling to suggest or one of the
positionals is `DoubleDash::Required` (clap's `last`) or `DoubleDash::Automatic`
(clap's `trailing_var_arg`).

So a spelling suggestion wins on an ordinary command — `--fore` against `--force`
is a typo, not a value somebody wanted forwarded — and a command whose whole job
is to collect a trailing command line says both, because there the flag really may
have been meant for what runs next. With nothing close enough to suggest, the
separator tip stands alone:

```
$ ex use --zzz
error: unexpected argument '--zzz' found

  tip: to pass '--zzz' as a value, use '-- --zzz'
```

Nothing is offered where the command takes no positional values at all, or once
the user has already typed a `--`: they know about the separator, so a second
refusal past it is a real mistake and repeating the advice they just took reads
as the CLI not listening.

## Where the separator state comes from

The parser tracks this as `separator_seen`, but `Error` does not carry it and
widening a public enum for a cold-path message is the wrong trade. `separator_before`
answers the same question from argv, using the offending token's own span to know
where to stop looking. One case is inexact: a `--` that a `DoubleDash::Preserve`
argument kept as a value counts as a separator when it should not. That can only
withhold the tip, never produce a wrong one.

## Behavior changes

Three existing assertions moved, and each was pinning the gap rather than a rule:

- `nothing_is_suggested_when_nothing_is_close` expected no tip at all for
  `ex use --zzz`. clap 4 prints the separator tip there; I ran it rather than
  recalling it.
- Two assertions spelled "no spelling suggestion here" as `!contains("tip:")`.
  They now say what they mean.
- `executable_views_emit_and_dispatch_from_argv0` counted occurrences of
  `--root-token` to check that an omitted host global is not suggested. The word
  is quoted back more than once now, so the count no longer expresses the intent;
  it asserts the absence of a spelling suggestion instead.

## Validation

New tests cover the `--`-only command (both tips, in clap's order), the ordinary
command beside it (spelling alone, so the pair above pins placement rather than
mere presence), a separator already given, and a command with no positionals. The
expected strings were checked against clap 4.6.6 built for the same shapes, not
written from memory. `cargo test --all --all-features` is green across 151 test
binaries, `cargo fmt` is clean, and `mise run render` produces no drift.

Found while fixing jdx/mise#13202, where a PowerShell quirk deletes the `--`
before mise ever sees it. That is not a usage bug — the separator never reaches
the parser — but this message is what a user gets when the separator is missing
for any reason, in any shell, including simply forgetting it.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

*AI-assisted — Tool: Claude Code; model: anthropic/claude-opus-5; version: 2.1.270.*

---

> [!NOTE]
> **Low Risk**
> User-visible diagnostic text only; behavior is aligned with clap and covered by expanded tests, with no parser or auth changes.
> 
> **Overview**
> Unexpected long-flag errors now follow **clap’s rule** for when to add a second tip: *to pass `'…'` as a value, use `'-- …'`*.
> 
> The tip appears when the active command accepts positional values, no `--` already precedes the refused token (inferred from argv via new `separator_before` / `value_capture` logic), and either there is no close spelling match or the command has a trailing/`--`-only positional (`DoubleDash::Required` or `Automatic`). Commands like **`exec`** that only take a command line after `--` can show **both** the “similar argument” and separator tips; ordinary commands with a near miss keep spelling only. Commands with no positionals, or tokens already after `--`, omit the separator tip.
> 
> Tests add an **`exec`** fixture and cover separator edge cases (literal tokens, repeated words, `--` placement). A facade test now asserts absence of spelling suggestions instead of counting token occurrences, since the new tip quotes the flag again.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 26e1dd28453c752e9635b0cd8e4640fd254db4ba. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved unexpected-option messages with clearer guidance for passing values that begin with `--`.
  - Corrected separator detection so later separators no longer suppress relevant value-passing guidance.
  - Preserved suppression of guidance when a separator appears before the value.
  - Refined similar-argument diagnostics to avoid displaying misleading spelling suggestions.
  - Improved handling of repeated arguments and commands that do not accept positional values, resulting in more accurate diagnostic tips.

### CODE REVIEW

File: argv/src/diagnostic.rs [hunk 1]
```diff
@@ -362,11 +422,36 @@ fn unexpected_flag(out: &mut String, style: Style, typed: &str, chain: &[&Comman
...
+    // clap's rule, transcribed rather than invented, down to the exception. Its comment:
+    // "`did_you_mean` is a lot more likely and should cause us to skip the `--` suggestion
+    // with the one exception being that the CLI is trying to capture arguments". So a
+    // misspelling wins on an ordinary command — `--fore` against `--force` is a typo, not a
+    // value somebody wanted forwarded — but a command that exists to collect a trailing
+    // command line says both, because there the flag really may have been meant for what runs
+    // next. `mise exec -- pnpm --version` is that shape.
+    let as_value = capture.takes_positionals
+        && !capture.separator_seen
+        && (near.is_empty() || capture.captures_trailing);
```
[the remark this answers is not in the document]
Comment (jdx) [reply argv/src/diagnostic.rs:459]: I ran this against the parser rather than reasoning about it, and the tip turns out to be unreachable in the case you describe. When a `Preserve` positional is the one waiting to be filled, an unknown flag never gets refused — it binds straight to that positional as a value, so no `UnknownFlag` is raised and no tip is rendered:

```
keep(preserve, var)  ["--zzz"]          -> KEPT=--zzz
keep(preserve, one)  ["--zzz"]          -> KEPT=--zzz
```

The tip only becomes reachable once the positional is *full*, and there `Preserve` behaves exactly like an ordinary positional — both error, and both still error after taking the advice:

```
keep(preserve, one)  ["x", "--zzz"]        -> KEPT=x, ERR UnexpectedArg
keep(preserve, one)  ["x", "--", "--zzz"]  -> KEPT=x, ERR UnexpectedArg
plain(optional, one) ["x", "--zzz"]        -> PLAIN=x, ERR UnexpectedArg
plain(optional, one) ["x", "--", "--zzz"]  -> PLAIN=x, ERR UnexpectedArg
```

So this is not a `Preserve` behavior — it is "every positional is already full", which clap has too: its condition is `has_positionals()`, not "an unfilled positional remains". That choice is deliberate and called out in `value_capture`'s doc comment; narrowing it would drop the tip on exactly the commands that fill their positionals greedily, which is where the mistake is easiest to make.

I also checked the combination that could have made `Preserve` genuinely different — a `Preserve` positional ahead of a `--`-only one, where a separator might be swallowed before it unlocks the trailing argument. It isn't:

```
mixed(preserve, required)  ["x", "--", "--zzz"]  -> KEPT=x, COMMAND=--zzz
```

Once the `Preserve` argument is full it stops claiming separators, so the advice works. No change made.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

*AI-assisted — Tool: Claude Code; model: anthropic/claude-opus-5; version: 2.1.270.*

File: argv/src/diagnostic.rs [hunk 2]
```diff
@@ -343,9 +343,69 @@ fn tip(style: Style, noun: &str, near: &[&str]) -> String {
...
 
+/// Whether a bare `--` stands in argv ahead of `token`.
+///
+/// The parser knows this as `separator_seen`, but the error it hands back does not carry the
+/// flag, and argv answers the same question without widening a public enum. A `--` that a
+/// `DoubleDash::Preserve` argument kept as a value counts here when it should not; that only
+/// ever withholds the tip below, which is the harmless direction to be wrong in.
+fn separator_before(argv: &[&std::ffi::OsStr], token: &[u8]) -> bool {
+    let limit = span_of_slice(argv, token)
+        .map(|span| span.index)
```
[the remark this answers is not in the document]
Comment (jdx) [reply argv/src/diagnostic.rs:363]: Correct, and CodeRabbit flagged the same thing independently — fixed in ef6b66fb2a.

Two changes, because the span lookup was wrong in both directions:

- `span_of_slice` matches on storage, so it only finds a token the parser sliced out of argv. It now falls back to matching the *word*, which locates an embedding's own token exactly rather than defaulting.
- When the token appears nowhere in argv, it no longer scans the whole line — it returns `false`. A token that is not in this argv belongs to a command line this argv does not describe, and answering "a separator precedes it" on the strength of a `--` standing somewhere unrelated is the one reading that is certainly wrong.

So the rule is now "suppress only where a `--` is shown to stand ahead of the refused word", instead of a guess in either direction.

Pinned both ways in `a_separator_later_on_the_line_does_not_count_as_already_given`: a separator *after* the refused word no longer suppresses, and one genuinely *before* it still does. Each half fails if the word match is removed, so the pair tests the lookup rather than the default.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

*AI-assisted — Tool: Claude Code; model: anthropic/claude-opus-5; version: 2.1.270.*

File: argv/src/diagnostic.rs [hunk 3]
```diff
@@ -343,9 +343,69 @@ fn tip(style: Style, noun: &str, near: &[&str]) -> String {
...
+        .unwrap_or(argv.len());
```
[the remark this answers is not in the document]
Comment (jdx) [reply argv/src/diagnostic.rs]: Correct, and CodeRabbit flagged the same thing independently — fixed in ef6b66fb2a.

Two changes, because the span lookup was wrong in both directions:

- `span_of_slice` matches on storage, so it only finds a token the parser sliced out of argv. It now falls back to matching the *word*, which locates an embedding's own token exactly rather than defaulting.
- When the token appears nowhere in argv, it no longer scans the whole line — it returns `false`. A token that is not in this argv belongs to a command line this argv does not describe, and answering "a separator precedes it" on the strength of a `--` standing somewhere unrelated is the one reading that is certainly wrong.

So the rule is now "suppress only where a `--` is shown to stand ahead of the refused word", instead of a guess in either direction.

Pinned both ways in `a_separator_later_on_the_line_does_not_count_as_already_given`: a separator *after* the refused word no longer suppresses, and one genuinely *before* it still does. Each half fails if the word match is removed, so the pair tests the lookup rather than the default.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

*AI-assisted — Tool: Claude Code; model: anthropic/claude-opus-5; version: 2.1.270.*

File: argv/src/diagnostic.rs [hunk 4]
```diff
@@ -343,9 +343,84 @@ fn tip(style: Style, noun: &str, near: &[&str]) -> String {
...
+    // take both. Matching the word covers the second kind; where the same word appears twice
+    // the first is taken, which is the only one a refusal can be about, since a `--` ahead of
+    // the later one would have stopped the parser treating it as a flag at all.
+    let Some(limit) = span_of_slice(argv, token)
+        .map(|span| span.index)
+        .or_else(|| {
+            argv.iter()
+                .position(|word| word.as_encoded_bytes() == token)
+        })
+    else {
```
[the remark this answers is not in the document]
Comment (jdx) [reply argv/src/diagnostic.rs:368]: You're right that my justification was wrong, and I've replaced it — the behavior it was defending turns out to be the one I'd want anyway.

The claim was that the first occurrence "is the only one a refusal can be about, since a `--` ahead of the later one would have stopped the parser treating it as a flag." That reasoning only holds for an error the parser produced, and the whole reason this fallback exists is errors it didn't. Fair catch.

What taking the first occurrence actually means is: **a separator counts as given only when one stands ahead of _every_ occurrence of the word** — because one ahead of the first is ahead of all of them. That is a question an embedding's error can answer, where "which occurrence did you mean" is not. Both readings are then defensible rather than arbitrary:

```
["use", "--zzz", "--", "--zzz"]   -> tip offered    (not every occurrence has one ahead)
["use", "--", "--zzz", "--zzz"]   -> tip suppressed (every one does, whichever was meant)
```

On the split case I kept the tip rather than dropping it. If the error meant the later `--zzz`, the user reads one line they can ignore; if it meant the earlier one, suppressing withholds the answer from someone who needs it. Of the two ways to be wrong on a cold error path, the redundant line is the smaller.

Pinned both rows in `a_repeated_word_is_answered_for_every_occurrence_of_it`, and the comment now states the rule instead of the bad argument. Fixed in 26e1dd2845.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

*AI-assisted — Tool: Claude Code; model: anthropic/claude-opus-5; version: 2.1.270.*
